### PR TITLE
Adding vast client use case examples

### PR DESCRIPTION
**Pull request by Rapha0511**
### Description

This PR add an example on how the vast_client can be implemented. 

### Type
- [ ] Breaking change
- [X] Enhancement
- [ ] Fix
- [ ] Documentation
- [ ] Tooling

### CODE REVIEW

File: examples/src/utils.js [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+  const creative = getCreative(validAd);
+  const mediaFileUrl = getMediaFileUrl(creative);
+  const variationData = getVariationData(validAd);
+  const companionCreative = getCreativeCompanion(validAd);
+
+  videoPlayer.src = mediaFileUrl;
+  videoPlayer.play();
+
+  // listen the the en ended event to stop the ad and remove trackers
```
Comment (clarissedmn) [inline examples/src/utils.js:73]:
```suggestion
  // listen the ended event to stop the ad and remove trackers
```

File: examples/src/utils.js [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+  videoPlayer.addEventListener('ended', (e) => {
+    stopAd();
+  });
+
+  //once we have the parsed vast we can instanciate the vastTracker
```
Comment (clarissedmn) [inline examples/src/utils.js:78]:
```suggestion
  //once we have the parsed vast we can instantiate the vastTracker
```

File: examples/src/utils.js [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+  }
+};
+
+/**
+ * play a preroll
+ */
+
+export const playPreroll = async () => {
+  const parsedVAST = await parsePreroll();
+  // here we a using the first ad from the parsed vast that contain creatives,
```
Comment (clarissedmn) [inline examples/src/utils.js:58]:
```suggestion
  // here we are using the first ad from the parsed vast that contain creatives,
```

File: examples/src/utils.js [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+/**
+ * @returns the image url of the variation to display
+ */
+
+const getVariationUrl = (variation) => {
+  if (!variation.staticResources.length) {
+    return null;
+  }
+  const variationImage = variation.staticResources[0].url;
```
Comment (clarissedmn) [inline examples/src/utils.js:302]:
```suggestion
  return variation.staticResources[0].url;
```

File: examples/src/utils.js [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+  return variationImage;
+};
+
+const displayVariation = (variationData) => {
+  const { imgurl, variation } = variationData;
```
Comment (clarissedmn) [inline examples/src/utils.js:307]:
```suggestion
  const { imgUrl, variation } = variationData;
```

File: examples/src/index.html [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  <head>
+    <meta charset="UTF-8" />
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <link rel="stylesheet" type="text/css" href="./styles.css" />
+    <title>Document</title>
+  </head>
+  <body>
+    <p>This page aim to demonstrate how to use the vast-client.js library and implement a simple preroll using a html5 video tag</p>
+    <div id="main">
```
Comment (clarissedmn) [inline examples/src/index.html:12]:
```suggestion
 <main role="main" id="main">
    <p>This page aim to demonstrate how to use the vast-client.js library and implement a simple preroll using a html5 video tag</p>
```

File: examples/src/utils.js [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+/**
+ * Get the first valid creative from an ad
+ */
+export const getCreative = (ad) => {
+  if (!ad.creatives) {
+    return null;
+  }
+  const validCreative = ad.creatives.find((creative) => creative.mediaFiles);
+  return validCreative || null;
+};
```
Comment (clarissedmn) [inline examples/src/utils.js:22]: I think it's more readable like this :) Idem for others. WDYT ?
```suggestion
export const getCreative = (ad) => {
  return ad.creatives?.find(creative => creative.mediaFiles) || null;
};
```

File: examples/src/utils.js [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+
+/**
+ * @returns a promise that resolves with the parsed VAST
+ */
+
+export const parsePreroll = async () => {
+  const VAST = 'https://statics.dmcdn.net/h/html/vast/simple-inline.xml';
+
+  let vastClient = new VASTClient();
```
Comment (clarissedmn) [inline examples/src/utils.js:44]: const ?

File: examples/src/utils.js [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+  const creative = getCreative(validAd);
+  const mediaFileUrl = getMediaFileUrl(creative);
+  const variationData = getVariationData(validAd);
+  const companionCreative = getCreativeCompanion(validAd);
+
+  videoPlayer.src = mediaFileUrl;
+  videoPlayer.play();
+
+  // listen the the en ended event to stop the ad and remove trackers
+  videoPlayer.addEventListener('ended', (e) => {
```
Comment (clarissedmn) [inline examples/src/utils.js:74]: Not sure, but can we do that instead ?
```suggestion
  videoPlayer.addEventListener('ended', stopAd)
```

File: examples/src/utils.js [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+
+/**
+ * Stop the ad and remove all EventListeners
+ */
+
+export const stopAd = () => {
+  videoPlayer.src = mainContent;
+  videoPlayer.load();
+  videoPlayer.play();
+  variationContainer.classList.add('hidden');
```
Comment (clarissedmn) [inline examples/src/utils.js:103]: Perhaps it's not necessary to add a classlist for this
```suggestion
variationContainer.style.display = "none";
```

File: examples/src/utils.js (hunk 7 above)
Comment (Rapha0511) [reply examples/src/utils.js:22]: yes definitly more readable !

File: examples/src/utils.js [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+};
+
+let prerollPlayed = false;
+
+videoPlayer.addEventListener('play', () => {
+  if (!prerollPlayed) {
+    playPreroll();
+    prerollPlayed = true;
+  }
+});
```
Comment (clementFrancon) [inline examples/src/utils.js:322]: Do we need this listener if it's already set in examples/src/index.js ?

**Review by clementFrancon — COMMENTED** [review]
Small comments otherwise LGTM, good work !

File: examples/src/utils.js (hunk 11 above)
Comment (Rapha0511) [reply examples/src/utils.js:322]: No, definitly something i forgot to remove :)

File: examples/src/index.html [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+          width="480"
+          height="360"
+          controls
+        ></video>
+      </div>
+      <p>Companion ad slot</p>
+      <div id ="variationContainer">
+        <img src="#" alt="" id="variationImg" />  
+      <div>
+    <main>
```
Comment (clarissedmn) [inline examples/src/index.html:26]:
```suggestion
      </div>
    </main>   
```

File: examples/src/index.html [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+          id="myVideo"
+          src="https://dmplayer.storage.googleapis.com/tech_test/midnight_sun_720p.mp4"
+          width="480"
+          height="360"
+          controls
+        ></video>
+      </div>
+      <p>Companion ad slot</p>
+      <div id ="variationContainer">
+        <img src="#" alt="" id="variationImg" />
```
Comment (clarissedmn) [inline examples/src/index.html:24]: This image is a informative one so we need to fill the alternative text :)

File: examples/src/index.html [hunk 14]
```diff
@@ -0,0 +1,29 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <link rel="stylesheet" type="text/css" href="./styles.css" />
+    <title>Document</title>
```
Comment (clarissedmn) [inline examples/src/index.html:8]: Something like that
```suggestion
   <title>Demonstrating Vast-client.js Library and Implementing Preroll</title>
```

File: examples/src/index.html
Comment (clarissedmn) [inline examples/src/index.html:1]: It's better for accessibility to include an H1 title :)

File: examples/src/utils.js [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+  videoPlayer.src = mediaFileUrl;
+  videoPlayer.play();
+
+  // listen the ended event to stop the ad and remove trackers
+  videoPlayer.addEventListener('ended', stopAd);
+
+  // once we have the parsed vast we can instantiate the vastTracker
+  vastTracker = new VASTTracker(null, validAd, creative);
+
+  //if we have a companion ad we instanciate the companionVastTracker
```
Comment (clarissedmn) [inline examples/src/utils.js:79]: same
```suggestion
  //if we have a companion ad we instantiate the companionVastTracker
```

File: examples/src/index.js [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+
+const videoPlayer = document.getElementById('myVideo');
+let prerollPlayed = false;
+
+videoPlayer.addEventListener('play', () => {
+  if (!prerollPlayed) {
+    playPreroll();
+    prerollPlayed = true;
+  }
+});
```
Comment (GuillaumeBAECHLER) [inline examples/src/index.js]: Maybe just:
```suggestion
videoPlayer.addEventListener('play', playPreroll, { once: true });
```

Comment (GuillaumeBAECHLER) [inline examples/src/index.js]: Maybe just:
```suggestion
videoPlayer.addEventListener('play', playPreroll, { once: true });
```

**Review by GuillaumeBAECHLER — COMMENTED** [review]
@Rapha0511 do you think we should add the package-lock file ?
And I think maybe .gitignore requires some changes otherwise when installing example app dependencies locally, the node_modules folder won't be ignored 😅

**Review by GuillaumeBAECHLER — APPROVED** [review]
LGTM!

**Pull request merged**